modules/product/src/Plugin/Field/FieldWidget/ProductVariationAttributesWidget.php

```diff
@@ -182,38 +172,40 @@ public function formElement(FieldItemListInterface $items, $delta, array $elemen
   }
 
   /**
-   * Selects a product variation from user input.
+   * Validation callback to set the variation value.
    *
-   * If there's no user input (form viewed for the first time), the default
-   * variation is returned.
+   * The attributes are used to resolve the variation value. This callback
+   * is used to ensure the variation value is updated to the currently
+   * selected variation via its attribute values.
    *
-   * @param \Drupal\commerce_product\Entity\ProductVariationInterface[] $variations
-   *   An array of product variations.
-   * @param array $user_input
-   *   The user input.
-   *
-   * @return \Drupal\commerce_product\Entity\ProductVariationInterface
-   *   The selected variation.
+   * @param array $form
+   *   The form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The form state.
    */
-  protected function selectVariationFromUserInput(array $variations, array $user_input) {
+  public static function setElementValue(array $form, FormStateInterface $form_state) {
+    $attribute_element = $form_state->getTriggeringElement();
+    $element = NestedArray::getValue($form, array_slice($attribute_element['#array_parents'], 0, -2));
+    $values = $form_state->getValue($element['#parents']);
+
+    /** @var \Drupal\commerce_product\Entity\ProductInterface $product */
+    $product = $form_state->get('product');
+    /** @var \Drupal\commerce_product\Entity\ProductVariationInterface[] $variations */
+    $variations = \Drupal::entityTypeManager()->getStorage('commerce_product_variation')->loadEnabled($product);
     $current_variation = reset($variations);
-    if (!empty($user_input)) {
-      $attributes = $user_input['attributes'];
-      foreach ($variations as $variation) {
-        $match = TRUE;
-        foreach ($attributes as $field_name => $value) {
-          if ($variation->getAttributeValueId($field_name) != $value) {
-            $match = FALSE;
-          }
-        }
-        if ($match) {
-          $current_variation = $variation;
-          break;
+    foreach ($variations as $variation) {
+      $match = TRUE;
+      foreach ($values['attributes'] as $field_name => $value) {
+        if ($variation->getAttributeValueId($field_name) != $value) {
+          $match = FALSE;
         }
       }
+      if ($match) {
+        $current_variation = $variation;
+        break;
+      }
     }
-
-    return $current_variation;
+    $form_state->setValueForElement($element['variation'], $current_variation->id());
```

Comment:
This does not persist. It only sets the value during the validation cycle, which causes test fails.

I think `commerce_product_variation_select` element is what we need. Handle loadEnabled, filtering access. Displaying attributes and resolving final variant ID. So we do not need to attempt to piece the attributes to a variation within the widget.